File: 23addLists.js

// iterative
// n = length of list 1
// m = length of list 2
// time:  O(max(n, m))
// space: O(max(n, m))
const addLists = (head1, head2) => {
  // todo
  let dummyHead = new Node(0);
  let tail = dummyHead;
  
  let carry = 0;
  let sum = 0;
  let digit;
  
  let current1 = head1;
  let current2 = head2;
  
  while (current1 || current2 || carry) {
    let val1 = current1 ? current1.val : 0;
    let val2 = current2 ? current2.val : 0;
    
    let sum = val1 + val2 + carry;
    let digit = sum % 10;
    if (sum > 9) carry = 1;
    else carry = 0;
    
    current1 = current1 ? current1.next : null;
    current2 = current2 ? current2.next : null;
    
    tail.next = new Node(digit);
    tail = tail.next;
  }
  
  return dummyHead.next;
};



// ------------------------------------------- //
// recursive
// n = length of list 1
// m = length of list 2
// time:  O(max(n, m))
// space: O(max(n, m))
const addLists = (head1, head2, carry=0) => {
  if (!head1 && !head2 && !carry) return null;
  
  let val1 = head1 ? head1.val : 0;
  let val2 = head2 ? head2.val : 0;
  let sum = val1 + val2 + carry;
  let nextCarry = sum > 9 ? 1 : 0;
  let digit = sum % 10;
  
  let next1 = head1 ? head1.next : null;
  let next2 = head2 ? head2.next : null;
  
  let newNode = new Node(digit);
  newNode.next = addLists(next1, next2, nextCarry);
  return newNode;
};

File: Python/23add_lists.py


class Node:
  def __init__(self, val):
    self.val = val
    self.next = None



# recursive
# n = length of list 1
# m = length of list 2
# Time: O(max(n, m))
# Space: O(max(n, m))

def add_lists(head_1, head_2, carry=0):
  if head_1 is None and head_2 is None and carry is 0:
    return None

  val1 = head_1.val if head_1 else 0
  val2 = head_2.val if head_2 else 0

  sum = val1 + val2 + carry
  digit = sum % 10
  carry = 1 if sum > 9 else 0

  next_1 = head_1.next if head_1 else head_1
  next_2 = head_2.next if head_2 else head_2

  new_node = Node(digit)
  new_node.next = add_lists(next_1, next_2, carry)

  return new_node



# iterative
# n = length of list 1
# m = length of list 2
# Time: O(max(n, m))
# Space: O(max(n, m))
def add_lists(head_1, head_2):
  new_node = Node('hello world. keep applyin')
  tail = new_node
  carry = 0

  while (head_1 is not None or head_2 is not None or carry == 1):
    val_1 = 0 if head_1 is None else head_1.val
    val_2 = 0 if head_2 is None else head_2.val

    sum = val_1 + val_2 + carry
    digit = sum % 10
    carry = 0 if sum <= 9 else 1

    tail.next = Node(digit)
    tail = tail.next

    if head_1:
      head_1 = head_1.next
    if head_2:
      head_2 = head_2.next

  return new_node.next
